lib/gitsh/git_command_list.rb

```diff
@@ -0,0 +1,62 @@
+module Gitsh
+  class GitCommandList
+    def initialize(env)
+      @env = env
+    end
+
+    def to_a
+      commands_from_list_cmds ||
+        commands_from_help('help -a --no-verbose') ||
+        commands_from_help('help -a') ||
+        []
+    end
+
+    private
+
+    attr_accessor :env
+
+    def commands_from_list_cmds
+      output = git_output('--list-cmds=main,nohelpers')
+
+      if output
+        output.lines.map(&:chomp).sort
+      else
+        false
+      end
+    end
+
+    def commands_from_help(command)
+      output = git_output(command)
+
+      if output
+        parse_help_output(output)
+      else
+        false
+      end
+    end
+
+    def parse_help_output(output)
+      output.
+        lines.
+        select { |line| line =~ /^  [a-z]/ }.
+        map { |line| line.split(/\s+/) }.
+        flatten.
+        reject { |cmd| cmd.empty? || cmd =~ /--/ }.
+        sort
+    end
+
+    def git_output(command)
+      output, _, status = Open3.capture3(git_command(command))
+
+      if status.success?
+        output.chomp
+      else
+        false
+      end
+    end
+
+    def git_command(sub_command)
```

Comment:
I'd like to state that I acknowledge that this is duplicated, have mild concerns about this, and am overall OK with it.

Reply:
Yep, that did happen. It also reminded me how much we pass a `Gitsh::Environment` instance around, which I'd love to get away from eventually.